### ISSUE TITLE
no-else-return rule should work in try-catch block

### ISSUE THREAD

**Issue by micnic** [eslint/eslint#11678]
**Tell us about your environment**

* **ESLint Version:** 5.16.0
* **Node Version:** 12.0.0
* **npm Version:** 6.9.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default
**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
    "env": {
        "es6": true,
        "node": true
    },
    "extends": "eslint:recommended",
    "rules":  {
        "no-else-return": "error"
    }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
class C {
    get(id) {
        if (this._options.get) {
            try {
                return this._options.get(id);
            } catch (error) {
                return Promise.reject(error);
            }
        } else { // Should warn about unnecessary else block
            return Promise.reject(new Error('Method not implemented'));
        }
    }
}
```

```bash
npx eslint test.js
```

**What did you expect to happen?**
It was expected to show an error that `else` block is unnecessary.

**What actually happened? Please include the actual, raw output from ESLint.**
No error displayed in the output, because there is a nested try-catch block inside the if block.

**Are you willing to submit a pull request to fix this bug?**
Not yet.

**Comment by kaicataldo** [comment]
This does seem like a bug to me and I was able to verify this using our demo page.

**Comment by g-plane** [comment]
I'm also able to reproduce this.

**Comment by mdjermanovic** [comment]
If this is a bug then the following might be considered as a bug, too?

```js
function test(x) {
  if (x) {
    switch (x) {
      case 1: return true;
      default: return false;
    }
  }
  else { // Should report unnecessary else?
    throw new Error('...');
  }
}
```

Or any other example where all paths inside the block will exit the function.

Perhaps the logic can be changed to check on `BlockStatement:exit` if the current `CodePathSegment` is unreachable? This would also cover `throw`, `break` and `continue` (when the `if` statement is in a loop) and other possible scenarios where `else` is unnecessary. I could try to implement that.

**Comment by micnic** [comment]
@mdjermanovic 

> Perhaps the logic can be changed to check on `BlockStatement:exit` if the current `CodePathSegment` is unreachable?

This is also what I was thinking, any code path that ends the execution of the function inside the `if` block should trigger the `no-else-return` rule. It would be very nice if you could implement it, I would try myself, but do not have much experience with ASTs and ESLint.

**Comment by mdjermanovic** [comment]
This could be solved with [Proposal: no-useless-else rule #8028](https://github.com/eslint/eslint/issues/8028) ?

The proposal is 'archived due to age', is it rejected for good? 

If there is still interest to include `no-useless-else` in ESLint rules, I could work on it.

**Comment by platinumazure** [comment]
@mdjermanovic The issue is also "accepted" so I think we could accept a pull request. Please feel free to work on it (please reach out if you aren't fully clear on the proposal to be implemented). Thanks!

**Comment by mdjermanovic** [comment]
Sure, I could try!

This will certainly need some more analysis. Should I open a new issue, or perhaps the #8028 should be re-opened?

**Comment by platinumazure** [comment]
@mdjermanovic Maybe best to open a new issue just to avoid any confusion around the previous conversation. Would you mind opening a new issue? Thanks!

**Comment by thewalla07** [comment]
I would like to work on this, could you please assign this issue to me?

**Comment by kaicataldo** [comment]
@thewalla07 Sounds great 🎉 ! You commenting here is sufficient enough. We use the assignees feature to indicate champions for issues from the team. Please feel free to comment here or stop by our [Gitter](https://gitter.im/eslint/eslint) if you need any assistance.

**Comment by anikethsaha** [comment]
I would like to take the `no-useless-else` if no one is working on it !

**Comment by mdjermanovic** [comment]
@eslint/eslint-team it would be nice to clarify what is expected here. 

This issue was originally accepted as a bug in `no-else-return`.

It can be also fixed by deprecating `no-else-return` in favor of a more generic rule `no-useless-else` (which would also cover `throw`, `break` and `continue` in addition to `return`). 

The new rule would use code path analysis and thus wouldn't have this bug.

Proposal for this rule was accepted in #8028, but that issue is now closed.

Can we confirm that a PR for `no-useless-else` will be accepted?

**Comment by anikethsaha** [comment]
It seems it is a code path analysis. So if this can be added in `no-unreachable` it would be much simpler though not sure about `BlockStatement:exit` 

[Check the rough implementation of `no-unreachable` (this should not be considered as `BlockStatement:exit` reports every block nodes )](https://astexplorer.net/#/gist/fbaf41975a0549e3631cc6fff4e10afe/c7042e4baeb669a503278b1542a7d10cc3578c88)


Output when adding `BlockStatement:exit`  to `no-unreachable`

```
   AssertionError [ERR_ASSERTION]: Should have 1 error but had 5: [ { ruleId: 'no-unreachable',
    severity: 1,
    message: 'Unreachable code.',
    line: 1,
    column: 15,
    nodeType: 'BlockStatement',
    messageId: 'unreachableCode',
    endLine: 9,
    endColumn: 11 },
  { ruleId: 'no-unreachable',
    severity: 1,
    message: 'Unreachable code.',
    line: 1,
    column: 40,
    nodeType: 'BlockStatement',
    messageId: 'unreachableCode',
    endLine: 7,
    endColumn: 10 },
  { ruleId: 'no-unreachable',
    severity: 1,
    message: 'Unreachable code.',
    line: 2,
    column: 17,
    nodeType: 'BlockStatement',
    messageId: 'unreachableCode',
    endLine: 4,
    endColumn: 14 },
  { ruleId: 'no-unreachable',
    severity: 1,
    message: 'Unreachable code.',
    line: 4,
    column: 29,
    nodeType: 'BlockStatement',
    messageId: 'unreachableCode',
    endLine: 6,
    endColumn: 14 },
  { ruleId: 'no-unreachable',
    severity: 1,
    message: 'Unreachable code.',
    line: 7,
    column: 16,
    nodeType: 'BlockStatement',
    messageId: 'unreachableCode',
    endLine: 9,
    endColumn: 10 } ]
      + expected - actual

      -5
      +1
      
      at testInvalidTemplate (lib/rule-tester/rule-tester.js:582:24)
      at Context.RuleTester.it (lib/rule-tester/rule-tester.js:764:25)
```



### May be like this ?

> this doesn't uses code path analysis

[Implementation using `IfStatement`](https://astexplorer.net/#/gist/857b9766b51f7bbde8307fec4baeee3f/aaf4bc12dcff5906674d34e45cff1911a95ee0f0)

**Comment by kaicataldo** [comment]
I'm still in favor of the proposal outlined in https://github.com/eslint/eslint/issues/8028. If someone is interested in championing that proposal, please feel free to create a new issue outlining where we landed in that issue.

**Comment by btmills** [comment]
Since this issue doesn't have a clear direction right now, we decided in today's TSC meeting to defer it from the v7.0.0 project so that we have time to get it right rather than pushing to get it out on the same timeline as the other changes in the release.

---

Speaking for myself now, I'm also in favor of the proposal from #8028 as a way to solve this.
